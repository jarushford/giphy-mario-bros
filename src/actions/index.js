export const setPlayers = (players) => ({
  type: 'SET_PLAYERS',
  players
})

export const clearPlayers = () => ({
  type: 'CLEAR_PLAYERS'
})

export const setError = (error) => ({
  type: 'SET_ERROR',
  error
})

export const clearError = () => ({
  type: 'CLEAR_ERROR'
})