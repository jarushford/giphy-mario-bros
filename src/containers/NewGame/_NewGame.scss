.final-screen {
  @include main();
  font-family: 'Mario';
}

.final-scores {
  font-size: 3rem;
  color: $white;
  text-shadow: $outlinethick;
  text-align: center;
  margin: 10px auto;
}

.scores-container {
  h3 {
    text-align: center;
    color: $white;
    text-shadow: $outlinethin;
    font-size: 1.8rem;
  }
}

.new-game {
  @include btn();
  margin-top: 30px;
}