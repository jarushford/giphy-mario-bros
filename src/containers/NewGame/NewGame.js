import React from 'react'
import { connect } from 'react-redux'
import { clearPlayers, clearChoices, setCurrentPlayer, setJudge, setUnusedIDs, resetRound, clearCaptions } from '../../actions';
import { Link } from 'react-router-dom'
import PropTypes from 'prop-types'
import { uid } from 'react-uid'

export function NewGame({ players, clearPlayers, clearChoices, setCurrentPlayer, setJudge, setUnusedIDs, resetRound, clearCaptions }) {
  const finalScores = players.map(player => {
    return (
      <h3 key={uid(player)}>Player {player.player}: {player.score}</h3>
    )
  })

  return (
    <section className="final-screen">
      <h1 className="final-scores">Final Scores</h1>
      <div className="scores-container">
        {finalScores}
      </div>
      <Link
        to='/select-players'
        onClick={() => {
          clearPlayers()
          clearChoices()
          setCurrentPlayer(2)
          setJudge()
          setUnusedIDs([])
          resetRound()
          clearCaptions()
        }}
      >
        <button className="new-game">New Game</button>
      </Link>
    </section>
  )
}

export const mapStateToProps = (state) => ({
  players: state.players
})

export const mapDispatchToProps = (dispatch) => ({
  clearPlayers: () => dispatch(clearPlayers()),
  clearChoices: () => dispatch(clearChoices()),
  setCurrentPlayer: (player) => dispatch(setCurrentPlayer(player)),
  setJudge: () => dispatch(setJudge()),
  setUnusedIDs: (IDs) => dispatch(setUnusedIDs(IDs)),
  resetRound: () => dispatch(resetRound()),
  clearCaptions: () => dispatch(clearCaptions())
})

export default connect(mapStateToProps, mapDispatchToProps)(NewGame)

NewGame.propTypes = {
  players: PropTypes.array.isRequired,
  clearPlayers: PropTypes.func.isRequired,
  clearChoices: PropTypes.func.isRequired,
  setCurrentPlayer: PropTypes.func.isRequired,
  setJudge: PropTypes.func.isRequired,
  setUnusedIDs: PropTypes.func.isRequired,
  resetRound: PropTypes.func.isRequired,
  clearCaptions: PropTypes.func.isRequired
}