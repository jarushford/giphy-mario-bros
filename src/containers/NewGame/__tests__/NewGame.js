import React from 'react'
import { NewGame, mapStateToProps, mapDispatchToProps } from '../NewGame'
import { clearPlayers, clearChoices, setCurrentPlayer, setJudge, setUnusedIDs } from '../../../actions'
import { shallow } from 'enzyme'

describe('NewGame', () => {
  let mockPlayers
  let mockClearPlayers
  let mockClearChoices
  let mockSetPlayer
  let mockSetJudge
  let mockSetIDs
  let wrapper

  beforeEach(() => {
    mockPlayers = [{ player: 1, score: 0 }, { player: 2, score: 0 }, { player: 3, score: 0 } ]
    mockClearPlayers = jest.fn()
    mockClearChoices = jest.fn()
    mockSetPlayer = jest.fn()
    mockSetJudge = jest.fn()
    mockSetIDs = jest.fn()
    wrapper = shallow(<NewGame 
      players={mockPlayers}
      clearPlayers={mockClearPlayers}
      clearChoices={mockClearChoices}
      setCurrentPlayer={mockSetPlayer}
      setJudge={mockSetJudge}
      setUnusedIDs={mockSetIDs}
    />)
  })

  describe('NewGame Component', () => {
    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should call clearPlayers when game is reset', () => {
      wrapper.find('Link').simulate('click')

      expect(mockClearPlayers).toBeCalled()
    })

    it('should call clearChoices when game is reset', () => {
      wrapper.find('Link').simulate('click')

      expect(mockClearChoices).toBeCalled()
    })

    it('should call setCurrentPlayer when game is reset', () => {
      wrapper.find('Link').simulate('click')

      expect(mockSetPlayer).toBeCalled()
    })

    it('should call setJudge when game is reset', () => {
      wrapper.find('Link').simulate('click')

      expect(mockSetJudge).toBeCalled()
    })

    it('should call setUnusedIDs when game is reset', () => {
      wrapper.find('Link').simulate('click')

      expect(mockSetIDs).toBeCalled()
    })
  })

  describe('mapStateToProps', () => {
    it('should return a props object with players', () => {
      const mockState = {
        players: mockPlayers,
        error: '',
        judge: 1
      }
      const expected = {
        players: mockPlayers
      }

      const mappedProps = mapStateToProps(mockState)

      expect(mappedProps).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {
    let mockDispatch

    beforeEach(() => {
      mockDispatch = jest.fn()
    })

    it('should dispatch clearPlayers when the prop clearPlayers is called', () => {
      const expected = clearPlayers()

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.clearPlayers()

      expect(mockDispatch).toBeCalledWith(expected)
    })

    it('should dispatch clearChoices when the prop clearPlayers is called', () => {
      const expected = clearChoices()

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.clearChoices()

      expect(mockDispatch).toBeCalledWith(expected)
    })

    it('should dispatch setCurrentPlayer when the prop clearPlayers is called', () => {
      const expected = setCurrentPlayer(2)

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.setCurrentPlayer(2)

      expect(mockDispatch).toBeCalledWith(expected)
    })

    it('should dispatch setJudge when the prop clearPlayers is called', () => {
      const expected = setJudge()

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.setJudge()

      expect(mockDispatch).toBeCalledWith(expected)
    })

    it('should dispatch setUnusedIDs when the prop clearPlayers is called', () => {
      const expected = setUnusedIDs([1, 2, 3])

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.setUnusedIDs([1, 2, 3])

      expect(mockDispatch).toBeCalledWith(expected)
    })
  })
})