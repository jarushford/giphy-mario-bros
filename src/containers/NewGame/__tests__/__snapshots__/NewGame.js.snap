// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewGame NewGame Component should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NewGame
    clearCaptions={[MockFunction]}
    clearChoices={[MockFunction]}
    clearPlayers={[MockFunction]}
    players={
      Array [
        Object {
          "player": 1,
          "score": 0,
        },
        Object {
          "player": 2,
          "score": 0,
        },
        Object {
          "player": 3,
          "score": 0,
        },
      ]
    }
    resetRound={[MockFunction]}
    setCurrentPlayer={[MockFunction]}
    setJudge={[MockFunction]}
    setUnusedIDs={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1
          className="final-scores"
        >
          Final Scores
        </h1>,
        <div
          className="scores-container"
        >
          <h3>
            Player 
            1
            : 
            0
          </h3>
          <h3>
            Player 
            2
            : 
            0
          </h3>
          <h3>
            Player 
            3
            : 
            0
          </h3>
        </div>,
        <Link
          onClick={[Function]}
          replace={false}
          to="/select-players"
        >
          <button
            className="new-game"
          >
            New Game
          </button>
        </Link>,
      ],
      "className": "final-screen",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Final Scores",
          "className": "final-scores",
        },
        "ref": null,
        "rendered": "Final Scores",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h3>
              Player 
              1
              : 
              0
            </h3>,
            <h3>
              Player 
              2
              : 
              0
            </h3>,
            <h3>
              Player 
              3
              : 
              0
            </h3>,
          ],
          "className": "scores-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "uid1",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Player ",
                1,
                ": ",
                0,
              ],
            },
            "ref": null,
            "rendered": Array [
              "Player ",
              1,
              ": ",
              0,
            ],
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": "uid2",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Player ",
                2,
                ": ",
                0,
              ],
            },
            "ref": null,
            "rendered": Array [
              "Player ",
              2,
              ": ",
              0,
            ],
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": "uid3",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Player ",
                3,
                ": ",
                0,
              ],
            },
            "ref": null,
            "rendered": Array [
              "Player ",
              3,
              ": ",
              0,
            ],
            "type": "h3",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <button
            className="new-game"
          >
            New Game
          </button>,
          "onClick": [Function],
          "replace": false,
          "to": "/select-players",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "New Game",
            "className": "new-game",
          },
          "ref": null,
          "rendered": "New Game",
          "type": "button",
        },
        "type": [Function],
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="final-scores"
          >
            Final Scores
          </h1>,
          <div
            className="scores-container"
          >
            <h3>
              Player 
              1
              : 
              0
            </h3>
            <h3>
              Player 
              2
              : 
              0
            </h3>
            <h3>
              Player 
              3
              : 
              0
            </h3>
          </div>,
          <Link
            onClick={[Function]}
            replace={false}
            to="/select-players"
          >
            <button
              className="new-game"
            >
              New Game
            </button>
          </Link>,
        ],
        "className": "final-screen",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Final Scores",
            "className": "final-scores",
          },
          "ref": null,
          "rendered": "Final Scores",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h3>
                Player 
                1
                : 
                0
              </h3>,
              <h3>
                Player 
                2
                : 
                0
              </h3>,
              <h3>
                Player 
                3
                : 
                0
              </h3>,
            ],
            "className": "scores-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "uid1",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Player ",
                  1,
                  ": ",
                  0,
                ],
              },
              "ref": null,
              "rendered": Array [
                "Player ",
                1,
                ": ",
                0,
              ],
              "type": "h3",
            },
            Object {
              "instance": null,
              "key": "uid2",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Player ",
                  2,
                  ": ",
                  0,
                ],
              },
              "ref": null,
              "rendered": Array [
                "Player ",
                2,
                ": ",
                0,
              ],
              "type": "h3",
            },
            Object {
              "instance": null,
              "key": "uid3",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Player ",
                  3,
                  ": ",
                  0,
                ],
              },
              "ref": null,
              "rendered": Array [
                "Player ",
                3,
                ": ",
                0,
              ],
              "type": "h3",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <button
              className="new-game"
            >
              New Game
            </button>,
            "onClick": [Function],
            "replace": false,
            "to": "/select-players",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "New Game",
              "className": "new-game",
            },
            "ref": null,
            "rendered": "New Game",
            "type": "button",
          },
          "type": [Function],
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
